compressors/fano_coder.py

```diff
@@ -0,0 +1,160 @@
+"""Fano Coding
+Shannon coding is a basic prefix free coding. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses Fano coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeTree, PrefixFreeEncoder, PrefixFreeDecoder, BinaryNode
+from core.prob_dist import ProbabilityDist
+
+
+class FanoTree(PrefixFreeTree):
+    def __init__(self, prob_dist):
+        super().__init__(prob_dist)
+        # sort the probability distribution
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        # initialize root node of Fano Tree
+        self.root_node = BinaryNode(id=None)
+
+        # build tree returns the root node
+        super().__init__(root_node=self.build_fano_tree(self.root_node, self.sorted_prob_dist))
+
+
+    @staticmethod
+    def criterion(dict_item):
+        """used for finding the symbol at which cumulative probability is nearest to 0.5
+        This symbol is used to split between left and right trees of Fano. We split right starting this symbol
+        to get most balanced tree.
+        """
+        key, value = dict_item
+        return abs(value - 0.5)
+
+    @staticmethod
+    def split_prob_dist_into_two(norm_sort_prob_dist):
+        """
+        Given a normalized and sorted probability distribution, split it into two sets of approximately equal
+        probabilities as described in Fano's algorithm.
+        """
+        # Get cumulative probability dict
+        cumulative_prob_dict = norm_sort_prob_dist.cumulative_prob_dict
+
+        # Split the tree into left and right nodes based on cumulative probability and recursively build
+        # FanoTree for these trees
+        recursive_prob_dicts = {
```

Comment:
not convinced about naming this recursive, because we don't do any recursion in the split function. One suggestion is `prob_dicts_after_split` or something similar

Reply:
+1, had discussed with pulkit to just have left, right.. as it makes things more clear ..